Feature: 808 edit application

aldipower opened the pull request:
This introduces an edit button in the configuration tab and version list.
The app modal will be opened with the app configuration prepopulated by the select app version.

I had to rewrite the ContainerSettingsComponent a little bit to prepopulate the docker container data in the form fields.
Now I can't add new container fields anymore. This is work to be done.

The EditAppModalComponent is reachable by a URL-query-param like this:
modal=editapp--%2Ffrontend%2Fgithub-pr-assigner%2Fdcos-ui--2015-07-21T17%3A14%3A37.058Z
What means ```:modalname--:appid--:appversion```

![edit-2](https://cloud.githubusercontent.com/assets/859154/9138792/65a385c2-3d26-11e5-8f9e-5f6c20ecd86c.png)
![edit-1](https://cloud.githubusercontent.com/assets/859154/9138795/679474ae-3d26-11e5-8fac-cc06d71b722b.png)
![edit-3](https://cloud.githubusercontent.com/assets/859154/9138796/69ba3430-3d26-11e5-86a6-f5b149cf08d4.png)
![edit-4](https://cloud.githubusercontent.com/assets/859154/9138798/6b68447a-3d26-11e5-9270-21b8f635bacc.png)

https://github.com/mesosphere/marathon/issues/808

We need a better more detailed error handling for the Docker Container Settings, but this is a bigger tasks what should be done separately:
https://github.com/mesosphere/marathon/issues/1985

aldipower commented:
@pierlo-upitup @philipnrmn This could be reviewed now, if you like.

pierlo-upitup commented on npm-shrinkwrap.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "marathon-ui",
-  "version": "0.10.0",
+  "version": "0.11.0",
```
I think this one slipped? Or do we want to update version number in this PR?

aldipower replied:
We already updated the version number in the package.json, but this is the first update of the npm-shrinkwrap.

pierlo-upitup replied:
OK, so does the shrinkwrap update belong here?

aldipower replied:
yes, shrinkwrap is updated in this PR

pierlo-upitup commented on src/css/main.less:
```diff
@@ -133,6 +133,10 @@ a,
 }
 
+.btn.pull-right {
+  margin-left: 5px;
```
I would have liked a comment here as to why this is necessary

aldipower replied:
So the buttons should have a gap between each other, if you have more then one..
https://cloud.githubusercontent.com/assets/859154/9138795/679474ae-3d26-11e5-8fac-cc06d71b722b.png
I think this is not comment-worth.

pierlo-upitup commented on src/js/actions/AppsActions.js:
```diff
@@ -133,7 +133,7 @@ var AppsActions = {
       });
   },
-  applySettingsOnApp: function (appId, settings) {
+  applySettingsOnApp: function (appId, settings, fromEdit = false) {
```
I suggest renaming `fromEdit` to `isEditing`. I like [the convention](https://github.com/airbnb/javascript#23.3) of using `is`/`has` + _-ing_ form to indicate boolean states.

aldipower replied:
:+1:

pierlo-upitup commented on src/js/components/AppVersionComponent.jsx:
```diff
@@ -33,6 +37,14 @@ var AppVersionComponent = React.createClass({
+    var appVersion = this.props.appVersion;
+    var router = this.context.router;
+    router.transitionTo(router.getCurrentPathname(), {}, {
+      modal: `editapp--${appVersion.id}--${appVersion.version}`
```
How about just `edit--` ? Alternatively, I'd prefer `edit-app--`

aldipower replied:
ok, let's stick with edit-app--

pierlo-upitup replied:
Oh I see. It all began with `newapp`. So maybe we should change that to `new-app` too...

aldipower replied:
Yeah, I will made the change to new-app in this PR.

pierlo-upitup replied:
Great!

pierlo-upitup commented on src/js/components/ContainerSettingsComponent.jsx:
```diff
@@ -14,11 +14,28 @@ var ContainerSettingsComponent = React.createClass({
   getInitialState: function () {
+    var container = this.props.model.container;
+
+    var [portMappingsCount, parametersCount, volumesCount] = Array(3).fill(1);
```
Cool!

aldipower replied:
;) ES6-power

mlunoe replied:
Hmm.. I think this is not readable, what is going on. What is wrong with being explicit?
```
var portMappingsCount = 1;
var parametersCount = 1;
var volumesCount = 1;
```

pierlo-upitup replied:
Yeah I like the ES6 power but in this case I agree it's overkill.

mlunoe replied:
Actually, why not have a configuration map? That way you can loop through it and make this thing much more dynamic. Say we wanted to add another field.

```
var config = {
  portMappings: {
    count: 1,
    getProperty: (obj) => obj.docker.portMappings.length || 1;
  },
  parameters: {
    count: 1,
    getProperty: (obj) => obj.parameters.length || 1;
  },
  volumes: {
    count: 1,
    getProperty: (obj) => obj.volumes.length || 1;
  }
};

if (container != null) {
  Object.keys(config).forEach(function (key) {
    var field = config[key];
    field.count = field.getProperty(container);
  });
}

// do something with rows as well
```

Or something similar. This way, we can just add to the configuration

aldipower replied:
@mlunoe I like this idea! Neat.

pierlo-upitup commented on src/js/components/ContainerSettingsComponent.jsx:
```diff
@@ -188,24 +224,32 @@ var ContainerSettingsComponent = React.createClass({
+    try {
+      network = model.container.docker.network;
+    } catch (e) {
+      network = "";
```
What if we did 

```
var network = "";
try {
   network = model.container.docker.network;
} catch(e) {};
```
(and same thing in `getVolumeRow()` above)
so that the default case is clearer to figure out?

aldipower replied:
Then eslint complains about the empty catch-block.. ;)

pierlo-upitup commented on src/js/components/FormGroupComponent.jsx:
```diff
@@ -1,4 +1,5 @@
 var classNames = require("classnames");
+var dotpathResolve = require("get-object-path");
```
Nitpicky little pick: how about `objectPath`? `dotpathResolve` also has the wrong camelCase (should be `dotPathResolve`)

aldipower replied:
Yeah, let's stick with objectPath.

pierlo-upitup commented on src/js/components/modals/AppModalComponent.jsx:
```diff
@@ -106,7 +121,7 @@ var NewAppModalComponent = React.createClass({
         new ValidationError(
           "general",
-          "App creation unsuccessful. Check your connection and try again."
+          "App creation unsuccessful. Check your app settings and try again."
```
Makes more sense!

pierlo-upitup commented on src/js/components/modals/AppModalComponent.jsx:
```diff
@@ -150,6 +165,25 @@ var NewAppModalComponent = React.createClass({
+      if ("docker" in container) {
+        if ("portMappings" in container.docker) {
+          container.docker.portMappings =
+            lazy(container.docker.portMappings).compact().value();
```
:+1:

pierlo-upitup commented:
Great work, @aldipower !

I noticed that you introduced `get-object-path` and because we are essentially re-implementing its `insert()` functionality with our own `util.serializedArrayToDictionary` I created an issue to rewrite those parts: https://github.com/mesosphere/marathon/issues/2027

Minor nitpicks aside, LGTM. @philipnrmn passing the ball to you for a final round

philipnrmn commented:
I noticed that when I attempt to edit an existing configuration, I can't change the docker image or the network mode - there seems to be some kind of re-render which overwrites anything I type. I can't find any other affected fields.

aldipower commented:
@philipnrmn The uneditable behaviour was the classic defaultValue thing:
https://facebook.github.io/react/docs/forms.html#default-value

I rebased @Poltergeist's PR into this and fixed the defaultValue could you take a look again?

Thanks!

mlunoe commented on src/js/components/ContainerSettingsComponent.jsx:
```diff
@@ -50,14 +67,18 @@ var ContainerSettingsComponent = React.createClass({
     return (
       <div key={`pm-${i}`} className="row duplicable-row">
         <div className="col-sm-3">
           <FormGroupComponent
```
I know this is not part of this PR, but just another suggestion:
These guys could be made dynamically as well.

aldipower replied:
True. :) Good point. Will create a refactor issue for that tomorrow.

aldipower commented:
@philipnrmn Could you take a look again?

philipnrmn commented:
There is now an issue with the app creation and edit modal. When attempting to add a docker image either via edit or via create, the modal attempts to post all of the empty fields, including Volumes. This results in a 400 error from the API.

aldipower commented:
@philipnrmn Good catch.
I fixed it again. Removed the already existing fix before by myself during the modifications of this PR... Hehe.

philipnrmn commented:
OK, thanks for all your hard work on this one! I'm going to merge now.

aldipower commented:
Cool, thanx for looking over it so carefully and detecting still some bugs.
